### PR TITLE
Enhancements and Bug Fixes for Game Logic in GameManager

### PR DESCRIPTION
**Pull request by thedovla**
This pull request introduces several enhancements and fixes to the Hanabi game server, including new game mechanics, improved logging, and expanded test coverage. The most significant changes are the addition of new methods for deck manipulation, enhanced handling of edge cases (e.g., empty deck, maximum strikes), and comprehensive unit tests to ensure robustness.

 Game mechanics and functionality:

Added `dealInitialCards` method in `GameManager` to handle the distribution of initial cards to players at the start of the game.
Enhanced strike handling in `GameState`: Automatically ends the game when the maximum number of strikes is reached and logs an error message.
Introduced new methods in `Deck` for better deck manipulation, including `clear`, `addCardToTop`, and `isEmpty`.

Logging improvements:

Improved logging for `HintAction` to include the indices of matching cards when a hint is given.
Enhanced logging in `PlayCardAction` to log the state of played cards and handle scenarios where the deck is empty. 

Bug Fixes:
Resolved an issue where the game could proceed after reaching the maximum number of strikes.
Fixed scenarios where invalid card indices in playCard would not return appropriate error messages.

Test coverage:

Added new unit tests in `GameManagerTest` to cover critical scenarios, including:
  - Discarding a card after the game is over.
  - Playing a card with an invalid index.
  - Handling an empty deck scenario.
  - Exhausting hint tokens and verifying failure cases.
  - Advancing turns and ensuring correct player rotation.
  - Initializing games with edge cases (minimum and maximum players).
  - Verifying game over behavior when maximum strikes are reached.

### CODE REVIEW

File: src/main/java/se2/server/hanabi/game/actions/PlayCardAction.java [hunk 1]
```diff
@@ -23,6 +23,10 @@ public ActionResult execute() {
             game.getLogger().warn("Attempt to play card after game over by player " + playerId);
             return ActionResult.failure("Game is already over");
         }
+        if (game.getDeck().isEmpty()) {
```
Comment (TJCur) [inline src/main/java/se2/server/hanabi/game/actions/PlayCardAction.java:26]: Players are still able to play even when the deck is empty.
one last round is playable

File: src/main/java/se2/server/hanabi/game/GameState.java [hunk 2]
```diff
@@ -169,8 +169,13 @@ public int getCurrentScore() {
      * Increment the strikes counter
      */
     public void incrementStrikes() {
+        logger.info("Incrementing strikes. Current strikes: " + strikes);
         strikes++;
         logger.warn("Strike count increased to " + strikes + " out of " + GameRules.MAX_STRIKES);
+        if (strikes >= GameRules.MAX_STRIKES) {
```
Comment (TJCur) [inline src/main/java/se2/server/hanabi/game/GameState.java:175]: Redundant check: after at the end of a action (Hint, Discard, Play) game.advanceTurn() is called. where gameState.checkEndCondition() is called. This is where the number of strikes is compare against the MAX_STRIKES.

File: src/main/java/se2/server/hanabi/game/actions/PlayCardAction.java [hunk 3]
```diff
@@ -51,22 +73,19 @@ public ActionResult execute() {
...
             }
 
-            game.drawCardToHand(playerId);
-            game.advanceTurn();
-            return ActionResult.success("You successfully played " + card);
-        } else {
-            game.getDiscardPile().add(card);
-            game.incrementStrikes();
-            if (game.getStrikes() >= GameRules.MAX_STRIKES) {
-                game.setGameOver(true);
```
Comment (TJCur) [inline src/main/java/se2/server/hanabi/game/actions/PlayCardAction.java:61]: This check occours in game.advanceturn(). see previous comment for incrementstrikes

File: src/main/java/se2/server/hanabi/game/actions/PlayCardAction.java [hunk 4]
```diff
@@ -51,22 +73,19 @@ public ActionResult execute() {
...
-                game.getLogger().error("Wrong card played by player " + playerId + ". Game over.");
-                return ActionResult.failure("Wrong card! Game over.");
+            if (game.getDeck().isEmpty()) {
```
Comment (TJCur) [inline src/main/java/se2/server/hanabi/game/actions/PlayCardAction.java:76]: this check occours already in game.drawCardToHans(playerId). line 82

**Review by TJCur — COMMENTED** [review]
Apart from the logging additions, I'm not sure if any of these changes are necessary. They seems to be duplicating existing logic.
